dbms/src/Core/Field.h

```diff
@@ -455,16 +456,24 @@ class Field
     template <typename T>
     void createConcrete(T && x)
     {
-        using JustT = std::decay_t<T>;
-        new (&storage) JustT(std::forward<T>(x));
-        which = TypeToEnum<JustT>::value;
+        using UnqualifiedType = std::decay_t<T>;
+        which = TypeToEnum<UnqualifiedType>::value;
+
+        // In both Field and PODArray, small types may be stored as wider types,
+        // e.g. char is stored as UInt64. Field can return this extended value
+        // with get<StorageType>(). To avoid uninitialized results from get(),
+        // we must initialize the entire wide stored type, and not just the
+        // nominal type.
+        using StorageType = NearestFieldType<UnqualifiedType>;
+        new (&storage) StorageType(std::forward<T>(x));
     }
 
     /// Assuming same types.
     template <typename T>
     void assignConcrete(T && x)
     {
         using JustT = std::decay_t<T>;
+        assert(which == TypeToEnum<JustT>::value);
```

Comment:
Doesn't this assert() in release build lead to perf degradation on Field operations?

Reply:
This is standard assert() from <cassert>, it is disabled in release builds (when NDEBUG is defined).